File: app/src/main/java/org/cru/godtools/service/AccountListRegistrationService.java
package org.cru.godtools.service;

import android.content.Context;
import android.os.AsyncTask;

import org.cru.godtools.api.GodToolsApi;
import org.cru.godtools.base.Settings;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.json.JSONObject;

import java.io.IOException;

import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;
import me.thekey.android.Attributes;
import me.thekey.android.TheKey;
import me.thekey.android.eventbus.event.LoginEvent;
import retrofit2.Response;
import timber.log.Timber;

import static org.cru.godtools.api.BuildConfig.CAMPAIGN_FORMS_ID;

public class AccountListRegistrationService {
    @NonNull
    private final GodToolsApi mApi;
    @NonNull
    private final TheKey mTheKey;
    @NonNull
    private final Settings mPrefs;

    private AccountListRegistrationService(@NonNull final Context context) {
        mApi = GodToolsApi.getInstance(context);
        mTheKey = TheKey.getInstance(context);
        mPrefs = Settings.getInstance(context);

        // register with EventBus
        EventBus.getDefault().register(this);

        // add the current user to Adobe Campaigns if they haven't been added already
        final String guid = mTheKey.getDefaultSessionGuid();
        if (guid != null && !mPrefs.isAddedToCampaign(guid)) {
            AsyncTask.THREAD_POOL_EXECUTOR.execute(() -> registerUser(mTheKey.getDefaultSessionGuid()));
        }
    }

    private static AccountListRegistrationService sInstance;

    public static AccountListRegistrationService start(@NonNull final Context context) {
        synchronized (AccountListRegistrationService.class) {
            if (sInstance == null) {
                sInstance = new AccountListRegistrationService(context.getApplicationContext());
            }

            return sInstance;
        }
    }

    @Subscribe(threadMode = ThreadMode.ASYNC)
    public void onLogin(@NonNull final LoginEvent event) {
        registerUser(event.getGuid());
    }

    @WorkerThread
    private void registerUser(@NonNull final String guid) {
        // short-circuit if we don't have valid attributes or the email is null
        final Attributes attrs = mTheKey.getAttributes(guid);
        if (!attrs.areValid() || attrs.getEmail() == null) {
            return;
        }

        // trigger the signup
        try {
            final Response<JSONObject> response = mApi.campaignForms
                    .signup(CAMPAIGN_FORMS_ID, attrs.getEmail(), attrs.getFirstName(), attrs.getLastName())
                    .execute();
            if (response.isSuccessful()) {
                mPrefs.setAddedToCampaign(guid, true);
            }
        } catch (IOException e) {
            Timber.tag("AccountListRegService")
                    .d(e, "error registering user in account list");
        }

        // TODO: if this failed due to no connectivity, we should retry when we have an internet connection
    }
}

File: app/src/main/kotlin/org/cru/godtools/service/AccountListRegistrationService.kt
package org.cru.godtools.service

import dagger.Lazy
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import org.cru.godtools.api.BuildConfig.CAMPAIGN_FORMS_ID
import org.cru.godtools.api.CampaignFormsApi
import org.cru.godtools.base.Settings
import org.cru.godtools.model.User
import org.cru.godtools.user.data.UserManager
import timber.log.Timber

@Singleton
class AccountListRegistrationService @Inject internal constructor(
    private val settings: Settings,
    userManager: UserManager,
    private val campaignFormsApi: Lazy<CampaignFormsApi>
) {
    private val coroutineScope = CoroutineScope(Dispatchers.Default)

    init {
        userManager.userFlow
            .onEach { it?.let { registerUser(it) } }
            .launchIn(coroutineScope)
    }

    private suspend fun registerUser(user: User) {
        val guid = user.ssoGuid
        val email = user.email ?: return
        if (settings.isAddedToCampaign(guid = guid)) return

        // trigger the signup
        try {
            val response = campaignFormsApi.get()
                .signup(CAMPAIGN_FORMS_ID, email, user.givenName, user.familyName)
            if (response.isSuccessful) settings.recordAddedToCampaign(guid = guid)
        } catch (e: IOException) {
            Timber.tag("AccountListRegService").d(e, "error registering user in account list")
        }

        // TODO: if this failed due to no connectivity, we should retry when we have an internet connection
    }
}
